features/332-sso/api.md

```diff
@@ -0,0 +1,311 @@
+# Overview
+
+The Auth gear supports single sign-on (SSO) via OAuth 2.0 and Custom Token.
+
+In the future we may support OpenID Connect.
+
+The only implemented OAuth 2.0 flow is [Authorization Code](https://tools.ietf.org/html/rfc6749#section-4.1) for devices that have access to a browser.
+
+In the future we may support [limited-input device](https://developers.google.com/identity/protocols/OAuth2ForDevices).
+
+In addition to the OAuth 2.0 Authorization Code flow, existing provider access tokens are also accepted. However, the API is disabled by default due to possible vulnerability. The details will be included in later section in this document.
+
+# OAuth 2.0 Authorization Code flow
+
+## Web browser with Popup
+
+The flow does not change the location of the current window, instead, it opens a new popup window of the provider authorization UI.
+
+### API
+
+```typescript
+interface LoginOptions {
+  // It is true, merge existing user. Default to true.
+  mergeExistingUser?: boolean;
+  // Which realm to consider during merging. Default to the "default" realm.
+  mergeRealm?: string;
+}
+
+interface PopupOptions {
+  // If provided, it overrides to the default scope.
+  // Note that the scope must be large enough to fetch user info.
+  // The default scope typically requests access to information
+  // like email.
+  scope?: string[];
```

Comment:
Also, this scope should be a server-side config.

Reply:
This is an escape hatch to allow the client to override the server config.

Reply:
I don't think we should allow client side to override scopes. What are the use cases that require customization of scopes at client side?

Reply:
The customization of scope is related to `getProviderProfiles`. If the developer can customize the scope, then `getProviderProfiles` will include data with respect to the requested scope. If we remove `getProviderProfiles`, then the developer should have no need to override the scope in the client side.

cc @chpapa

Reply:
I tends to think it needs to be a parameters over configuration.

If the concern is server side or client side, we can make function (`getProviderProfiles`) with this parameter to be a server side call instead of Client side APIs.

(BTW, in Skygear v1, at client side we have JS/iOS/Android SDK, and part of the JS SDK intended for use at server side with Cloud Functions only; At Skygear v2, I remember there were decisions iOS/Android/JS SDK would be for client side; And server side functions would be as a REST API so we can be programming language agnostic on server side)

Reply:
I'm a little bit concern if our current spec are clear enough in specifiying which APIs are client side SDK, which is server side... @louischan-oursky @kiootic